examples/kmeans/kmeans.py

```diff
@@ -0,0 +1,88 @@
+import six
+try:
+    import matplotlib
+    matplotlib.use('Agg')
+    import matplotlib.pyplot as plt
+except ImportError:
+    pass
+
+import numpy as np
+
+import cupy
+
+
+def fit(X, n_clusters, max_iter, elem):
+    assert X.ndim == 2
+    xp = cupy.get_array_module(X)
+    pred = xp.zeros(len(X), dtype=np.int32)
+    initial_indexes = np.random.choice(len(X), n_clusters,
+                                       replace=False).astype(np.int32)
+    centers = X[initial_indexes]
+    data_num = X.shape[0]
+    data_dim = X.shape[1]
+
+    for _ in six.moves.range(max_iter):
+        # calculate distances and label
+        distances = xp.zeros((data_num, n_clusters), dtype=np.float32)
+        if elem is False or xp == np:
+            distances = xp.linalg.norm(X[:, None, :] - centers[None, :, :],
+                                       axis=2)
+        else:
+            cupy.ElementwiseKernel(
+                'S data, raw S centers, int32 dim', 'raw S dist',
+                '''
+                int cent_ind1[] = {0, i % dim};
+                int cent_ind2[] = {1, i % dim};
+                int dist_ind1[] = {i / dim, 0};
+                int dist_ind2[] = {i / dim, 1};
+                double diff1 = centers[cent_ind1] - data;
+                double diff2 = centers[cent_ind2] - data;
+                atomicAdd(&dist[dist_ind1], diff1 * diff1);
+                atomicAdd(&dist[dist_ind2], diff2 * diff2);
+                ''',
+                'calc_distances'
+            )(X, centers, data_dim, distances)
+
+        new_pred = xp.argmin(distances, axis=1).astype(np.int32)
+        if xp.all(new_pred == pred):
+            break
+        pred = new_pred
+
+        # calculate centers
+        if elem is False or xp == np:
+            centers = xp.empty((0, data_dim), dtype=np.float32)
+            for i in range(n_clusters):
+                centers = xp.vstack((centers,
```

Comment:
Don't call vstack recursively. It is very slow. 
`centers = xp.stack([X[pred == i].mean(axis=0) for i in six.moves.range(n_clusters)])`